File: tosctl/src/validator-auth/src/transport.rs
//! Strict framing and correlation. Parsed frames do not establish authority.
use crate::{
    api_types::validate_api_binary,
    codec::{decode, Error, Hash},
    crypto::digest,
    types::{ApiError, ResultRequest, SignRequest},
};
use std::collections::BTreeMap;
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TransportFrame {
    pub method: u8,
    pub request_id: Hash,
    pub payload: Vec<u8>,
    pub response: bool,
    pub error: bool,
}
struct Json<'a> {
    raw: &'a [u8],
    pos: usize,
}
impl<'a> Json<'a> {
    fn space(&mut self) {
        while self.pos < self.raw.len()
            && matches!(self.raw[self.pos], b' ' | b'\t' | b'\r' | b'\n')
        {
            self.pos += 1;
        }
    }
    fn take(&mut self, c: u8) -> bool {
        self.space();
        if self.raw.get(self.pos) == Some(&c) {
            self.pos += 1;
            true
        } else {
            false
        }
    }
    fn string(&mut self) -> Result<String, Error> {
        if !self.take(b'"') {
            return Err(Error("json-syntax"));
        }
        let mut out = String::new();
        while let Some(&byte) = self.raw.get(self.pos) {
            self.pos += 1;
            let mut c = byte;
            if c == b'"' {
                return Ok(out);
            }
            if !(32..128).contains(&c) {
                return Err(Error("transport-string"));
            }
            if c == b'\\' {
                let esc = *self.raw.get(self.pos).ok_or(Error("json-syntax"))?;
                self.pos += 1;
                c = match esc {
                    b'"' => b'"',
                    b'\\' => b'\\',
                    b'/' => b'/',
                    b'b' => 8,
                    b'f' => 12,
                    b'n' => 10,
                    b'r' => 13,
                    b't' => 9,
                    b'u' => {
                        let end = self.pos.checked_add(4).ok_or(Error("json-syntax"))?;
                        let digits = self.raw.get(self.pos..end).ok_or(Error("json-syntax"))?;
                        let mut value = 0u16;
                        for &digit in digits {
                            let n = match digit {
                                b'0'..=b'9' => digit - b'0',
                                b'a'..=b'f' => digit - b'a' + 10,
                                b'A'..=b'F' => digit - b'A' + 10,
                                _ => return Err(Error("json-syntax")),
                            };
                            value = value * 16 + u16::from(n);
                        }
                        self.pos = end;
                        if value >= 128 {
                            return Err(Error("transport-string"));
                        }
                        value as u8
                    }
                    _ => return Err(Error("json-syntax")),
                };
            }
            out.push(char::from(c));
        }
        Err(Error("json-syntax"))
    }
    fn parse(&mut self) -> Result<BTreeMap<String, Option<String>>, Error> {
        let mut fields = BTreeMap::new();
        if !self.take(b'{') {
            return Err(Error("json-syntax"));
        }
        if !self.take(b'}') {
            loop {
                let key = self.string()?;
                if fields.contains_key(&key) {
                    return Err(Error("duplicate-json-key"));
                }
                if fields.len() >= 4 {
                    return Err(Error("transport-fields"));
                }
                if !self.take(b':') {
                    return Err(Error("json-syntax"));
                }
                self.space();
                let value = if self.raw.get(self.pos..self.pos + 4) == Some(b"null") {
                    self.pos += 4;
                    None
                } else {
                    Some(self.string()?)
                };
                fields.insert(key, value);
                if self.take(b'}') {
                    break;
                }
                if !self.take(b',') {
                    return Err(Error("json-syntax"));
                }
            }
        }
        self.space();
        if self.pos != self.raw.len() {
            return Err(Error("json-syntax"));
        }
        Ok(fields)
    }
}
fn unhex(value: Option<&str>, limit: usize) -> Result<Vec<u8>, Error> {
    let value = value.ok_or(Error("hex"))?;
    if value.len() % 2 != 0 || value.len() / 2 > limit {
        return Err(Error("hex"));
    }
    let mut out = Vec::new();
    out.try_reserve(value.len() / 2).map_err(|_| Error("allocation"))?;
    for pair in value.as_bytes().chunks_exact(2) {
        let mut byte = 0;
        for &c in pair {
            byte = byte * 16
                + match c {
                    b'0'..=b'9' => c - b'0',
                    b'a'..=b'f' => c - b'a' + 10,
                    _ => return Err(Error("hex")),
                };
        }
        out.push(byte);
    }
    Ok(out)
}
fn hex(bytes: &[u8]) -> String {
    const DIGITS: &[u8] = b"0123456789abcdef";
    let mut out = String::with_capacity(bytes.len() * 2);
    for &b in bytes {
        out.push(char::from(DIGITS[usize::from(b >> 4)]));
        out.push(char::from(DIGITS[usize::from(b & 15)]));
    }
    out
}
fn request_id(method: u8, raw: &[u8]) -> Result<Hash, Error> {
    match method {
        1 => Ok([0; 32]),
        5 => Ok(decode::<SignRequest>(raw)?.request_id),
        6 => Ok(decode::<ResultRequest>(raw)?.request_id),
        _ => {
            let mut bytes = vec![method];
            bytes.extend_from_slice(raw);
            digest("api-request", &bytes)
        }
    }
}
fn validate_frame(frame: &TransportFrame) -> Result<(), Error> {
    if frame.error && !frame.response {
        return Err(Error("result-or-error"));
    }
    validate_api_binary(frame.method, frame.response, frame.error, &frame.payload)?;
    if !frame.response && request_id(frame.method, &frame.payload)? != frame.request_id {
        return Err(Error("request-correlation"));
    }
    if frame.error {
        let e = decode::<ApiError>(&frame.payload)?;
        if e.request_id != frame.request_id || e.method != frame.method {
            return Err(Error("error-correlation"));
        }
        if !(1..=14).contains(&e.code) || e.request_state > 4 {
            return Err(Error("error-code"));
        }
        let read = matches!(frame.method, 1 | 2 | 6 | 8..=15);
        if e.retryable != u8::from(read && (10..=12).contains(&e.code)) {
            return Err(Error("error-retry"));
        }
        std::str::from_utf8(&e.message).map_err(|_| Error("error-message"))?;
    }
    Ok(())
}
pub fn decode_transport_frame(
    raw: &[u8],
    method: u8,
    response: bool,
    expected: Option<&Hash>,
) -> Result<TransportFrame, Error> {
    if raw.len() > 4_194_304 {
        return Err(Error("transport-bound"));
    }
    let fields = Json { raw, pos: 0 }.parse()?;
    let required: &[&str] = if response {
        &["api_version", "request_id", "result", "error"]
    } else {
        &["api_version", "request_id", "request"]
    };
    if fields.len() != required.len()
        || required.iter().any(|k| !fields.contains_key(*k))
        || fields.get("api_version").and_then(|x| x.as_deref()) != Some("1")
    {
        return Err(Error("transport-fields"));
    }
    let field = |key: &str| fields.get(key).and_then(|x| x.as_deref());
    let request_id: Hash =
        unhex(field("request_id"), 32)?.try_into().map_err(|_| Error("hex-width"))?;
    if expected.is_some_and(|id| id != &request_id) {
        return Err(Error("response-correlation"));
    }
    if response && field("result").is_some() == field("error").is_some() {
        return Err(Error("result-or-error"));
    }
    let error = response && field("error").is_some();
    let payload = unhex(
        field(if response {
            if error {
                "error"
            } else {
                "result"
            }
        } else {
            "request"
        }),
        2_000_000,
    )?;
    let frame = TransportFrame { method, request_id, payload, response, error };
    validate_frame(&frame)?;
    Ok(frame)
}
pub fn encode_transport_frame(frame: &TransportFrame) -> Result<String, Error> {
    validate_frame(frame)?;
    let mut out = format!("{{\"api_version\":\"1\",\"request_id\":\"{}\",", hex(&frame.request_id));
    out.push_str(if frame.response {
        if frame.error {
            "\"result\":null,\"error\":\""
        } else {
            "\"error\":null,\"result\":\""
        }
    } else {
        "\"request\":\""
    });
    out.push_str(&hex(&frame.payload));
    out.push_str("\"}");
    if out.len() > 4_194_304 {
        return Err(Error("transport-bound"));
    }
    Ok(out)
}

File: validator/auth/transport.cpp
#include <map>
#include <optional>

#include "api-types.h"
#include "transport.h"
namespace tos::auth {
namespace {
using Fields = std::map<std::string, std::optional<std::string>>;
int digit(char c) {
  return c >= '0' && c <= '9' ? c - '0' : c >= 'a' && c <= 'f' ? c - 'a' + 10 : -1;
}
int escape_digit(char c) {
  return c >= 'A' && c <= 'F' ? c - 'A' + 10 : digit(c);
}
class Json {
  std::string_view raw;
  std::size_t pos = 0;
  void space() {
    while (pos < raw.size() && (raw[pos] == ' ' || raw[pos] == '\n' || raw[pos] == '\r' || raw[pos] == '\t'))
      ++pos;
  }
  bool take(char c) {
    space();
    if (pos == raw.size() || raw[pos] != c)
      return false;
    ++pos;
    return true;
  }
  Result<std::string> string() {
    if (!take('"'))
      return Error{"json-syntax"};
    std::string out;
    while (pos < raw.size()) {
      unsigned char c = raw[pos++];
      if (c == '"')
        return out;
      if (c < 32 || c >= 128)
        return Error{"transport-string"};
      if (c == '\\') {
        if (pos == raw.size())
          return Error{"json-syntax"};
        char esc = raw[pos++];
        switch (esc) {
          case '"':
            c = '"';
            break;
          case '\\':
            c = '\\';
            break;
          case '/':
            c = '/';
            break;
          case 'b':
            c = '\b';
            break;
          case 'f':
            c = '\f';
            break;
          case 'n':
            c = '\n';
            break;
          case 'r':
            c = '\r';
            break;
          case 't':
            c = '\t';
            break;
          case 'u': {
            if (raw.size() - pos < 4)
              return Error{"json-syntax"};
            unsigned value = 0;
            for (int i = 0; i < 4; ++i) {
              int n = escape_digit(raw[pos++]);
              if (n < 0)
                return Error{"json-syntax"};
              value = value * 16 + unsigned(n);
            }
            // Every accepted transport name and value is ASCII. Other Unicode
            // cannot name a valid field or encode the version/hex payload.
            if (value >= 128)
              return Error{"transport-string"};
            c = static_cast<unsigned char>(value);
            break;
          }
          default:
            return Error{"json-syntax"};
        }
      }
      out.push_back(static_cast<char>(c));
    }
    return Error{"json-syntax"};
  }

 public:
  explicit Json(std::string_view input) : raw(input) {
  }
  Result<Fields> parse() {
    Fields fields;
    if (!take('{'))
      return Error{"json-syntax"};
    if (take('}')) {
      space();
      if (pos != raw.size())
        return Error{"json-syntax"};
      return fields;
    }
    for (;;) {
      auto key = string();
      if (!key.ok())
        return key.error();
      if (fields.contains(key.value()))
        return Error{"duplicate-json-key"};
      if (fields.size() >= 4)
        return Error{"transport-fields"};
      if (!take(':'))
        return Error{"json-syntax"};
      space();
      std::optional<std::string> value;
      if (raw.substr(pos, 4) == "null")
        pos += 4;
      else {
        auto s = string();
        if (!s.ok())
          return s.error();
        value = std::move(s.value());
      }
      fields.emplace(std::move(key.value()), std::move(value));
      if (take('}'))
        break;
      if (!take(','))
        return Error{"json-syntax"};
    }
    space();
    if (pos != raw.size())
      return Error{"json-syntax"};
    return fields;
  }
};
Result<Bytes> unhex(const std::optional<std::string>& text, std::size_t limit) {
  if (!text || text->size() % 2 || text->size() / 2 > limit)
    return Error{"hex"};
  Bytes out;
  out.reserve(text->size() / 2);
  for (std::size_t i = 0; i < text->size(); i += 2) {
    int a = digit((*text)[i]), b = digit((*text)[i + 1]);
    if (a < 0 || b < 0)
      return Error{"hex"};
    out.push_back(static_cast<std::uint8_t>(a * 16 + b));
  }
  return out;
}
std::string hex(std::span<const std::uint8_t> raw) {
  constexpr char chars[] = "0123456789abcdef";
  std::string out;
  out.reserve(raw.size() * 2);
  for (auto c : raw) {
    out.push_back(chars[c >> 4]);
    out.push_back(chars[c & 15]);
  }
  return out;
}
bool utf8(std::span<const std::uint8_t> raw) {
  for (std::size_t i = 0; i < raw.size();) {
    auto c = raw[i++];
    if (c < 128)
      continue;
    unsigned count, value, min;
    if (c >= 0xc2 && c <= 0xdf) {
      count = 1;
      value = c & 31;
      min = 0x80;
    } else if (c >= 0xe0 && c <= 0xef) {
      count = 2;
      value = c & 15;
      min = 0x800;
    } else if (c >= 0xf0 && c <= 0xf4) {
      count = 3;
      value = c & 7;
      min = 0x10000;
    } else
      return false;
    if (count > raw.size() - i)
      return false;
    while (count--) {
      auto x = raw[i++];
      if ((x & 0xc0) != 0x80)
        return false;
      value = (value << 6) | (x & 63);
    }
    if (value < min || value > 0x10ffff || (value >= 0xd800 && value <= 0xdfff))
      return false;
  }
  return true;
}
Result<Hash> request_id(std::uint8_t method, std::span<const std::uint8_t> bytes) {
  if (method == 1)
    return Hash{};
  if (method == 5) {
    auto v = decode<SignRequest>(bytes);
    if (!v.ok())
      return v.error();
    return v.value().request_id_;
  }
  if (method == 6) {
    auto v = decode<ResultRequest>(bytes);
    if (!v.ok())
      return v.error();
    return v.value().request_id_;
  }
  Bytes input{method};
  input.insert(input.end(), bytes.begin(), bytes.end());
  return digest("api-request", input);
}
Result<bool> validate_frame(const TransportFrame& frame) {
  if (frame.error && !frame.response)
    return Error{"result-or-error"};
  auto valid = validate_api_binary(frame.method, frame.response, frame.error, frame.payload);
  if (!valid.ok())
    return valid.error();
  if (!frame.response) {
    auto id = request_id(frame.method, frame.payload);
    if (!id.ok())
      return id.error();
    if (id.value() != frame.request_id)
      return Error{"request-correlation"};
  }
  if (frame.error) {
    auto v = decode<ApiError>(frame.payload);
    if (!v.ok())
      return v.error();
    const auto& error = v.value();
    if (error.request_id_ != frame.request_id || error.method_ != frame.method)
      return Error{"error-correlation"};
    if (error.code_ < 1 || error.code_ > 14 || error.request_state_ > 4)
      return Error{"error-code"};
    bool read = frame.method == 1 || frame.method == 2 || frame.method == 6 || frame.method >= 8;
    if (error.retryable_ != static_cast<unsigned>(read && error.code_ >= 10 && error.code_ <= 12))
      return Error{"error-retry"};
    if (!utf8(error.message_))
      return Error{"error-message"};
  }
  return true;
}
}  // namespace
Result<TransportFrame> decode_transport_frame(std::string_view json, std::uint8_t method, bool response,
                                              const Hash* expected) {
  if (json.size() > 4194304)
    return Error{"transport-bound"};
  auto parsed = Json(json).parse();
  if (!parsed.ok())
    return parsed.error();
  const auto& fields = parsed.value();
  if (fields.size() != (response ? 4 : 3) || !fields.contains("api_version") ||
      fields.at("api_version") != std::optional<std::string>("1") || !fields.contains("request_id") ||
      (response ? (!fields.contains("result") || !fields.contains("error")) : !fields.contains("request")))
    return Error{"transport-fields"};
  auto id = unhex(fields.at("request_id"), 32);
  if (!id.ok())
    return id.error();
  if (id.value().size() != 32)
    return Error{"hex-width"};
  TransportFrame frame;
  frame.method = method;
  frame.response = response;
  std::copy(id.value().begin(), id.value().end(), frame.request_id.begin());
  if (expected && *expected != frame.request_id)
    return Error{"response-correlation"};
  if (response) {
    if (fields.at("result").has_value() == fields.at("error").has_value())
      return Error{"result-or-error"};
    frame.error = fields.at("error").has_value();
  }
  auto payload = unhex(fields.at(response ? (frame.error ? "error" : "result") : "request"), 2000000);
  if (!payload.ok())
    return payload.error();
  frame.payload = std::move(payload.value());
  auto valid = validate_frame(frame);
  if (!valid.ok())
    return valid.error();
  return frame;
}
Result<std::string> encode_transport_frame(const TransportFrame& frame) {
  auto valid = validate_frame(frame);
  if (!valid.ok())
    return valid.error();
  std::string result = "{\"api_version\":\"1\",\"request_id\":\"" + hex(frame.request_id) + "\",";
  if (frame.response)
    result += frame.error ? "\"result\":null,\"error\":\"" : "\"error\":null,\"result\":\"";
  else
    result += "\"request\":\"";
  result += hex(frame.payload);
  result += "\"}";
  if (result.size() > 4194304)
    return Error{"transport-bound"};
  return result;
}
}  // namespace tos::auth
